metabonding/src/claim_progress.rs

```diff
@@ -0,0 +1,286 @@
+        self.claim_flags[index] = CLAIMED;
+    }
+}
+
+#[elrond_wasm::module]
+pub trait ClaimProgressModule {
+    fn get_claimable_weeks(
+        &self,
+        tracker: &dyn ClaimProgressTracker,
+        start_week: Week,
+        end_week: Week,
+    ) -> ManagedVec<Week> {
+        let mut claimable_weeks = ManagedVec::new();
+        for week in start_week..=end_week {
+            if tracker.can_claim_for_week(week) {
+                claimable_weeks.push(week);
+            }
+        }
+
+        claimable_weeks
+    }
+
+    fn get_grace_weeks_progress(
+        &self,
+        user: &ManagedAddress,
+        nr_first_grace_weeks: Week,
+        current_week: Week,
+    ) -> ClaimProgressGraceWeeks<Self::Api> {
+        if current_week > nr_first_grace_weeks {
+            return ClaimProgressGraceWeeks::new(ManagedVec::new());
+        }
+
+        let mapper = self.claim_progress_grace_weeks(user);
+        if !mapper.is_empty() {
+            return mapper.get();
+        }
+
+        // index 0 is unused
+        let mut claim_flags = ManagedVec::from_single_item(NOT_CLAIMED);
+        for week in FIRST_WEEK..=nr_first_grace_weeks {
+            let old_flag = self.legacy_rewards_claimed_flag(user, week).get();
+            claim_flags.push(old_flag);
+        }
+
+        ClaimProgressGraceWeeks::new(claim_flags)
+    }
+
+    fn get_shifting_progress(
+        &self,
+        user: &ManagedAddress,
+        current_week: Week,
+    ) -> ShiftingClaimProgress {
+        let mapper = self.shifting_claim_progress(user);
+        if !mapper.is_empty() {
+            let mut existing_progress = mapper.get();
+            existing_progress.shift_if_needed(current_week);
+
+            return existing_progress;
+        }
+
+        let mut claim_flags = DEFAULT_CLAIM_FLAGS;
+        let first_accepted_week =
+            ShiftingClaimProgress::get_first_index_week_for_new_current_week(current_week);
+        for (i, week) in (first_accepted_week..=current_week).enumerate() {
+            let old_flag = self.legacy_rewards_claimed_flag(user, week).get();
+            claim_flags[i] = old_flag;
+        }
+
+        ShiftingClaimProgress::new(claim_flags, current_week)
+    }
+
+    #[storage_mapper("rewardsNrFirstGraceWeeks")]
+    fn rewards_nr_first_grace_weeks(&self) -> SingleValueMapper<Week>;
+
+    #[storage_mapper("claimProgressGraceWeeks")]
+    fn claim_progress_grace_weeks(
+        &self,
+        user: &ManagedAddress,
+    ) -> SingleValueMapper<ClaimProgressGraceWeeks<Self::Api>>;
+
+    #[storage_mapper("shiftingClaimProgress")]
+    fn shifting_claim_progress(
+        &self,
+        user: &ManagedAddress,
+    ) -> SingleValueMapper<ShiftingClaimProgress>;
+
+    #[storage_mapper("rewardsClaimed")]
+    fn legacy_rewards_claimed_flag(
+        &self,
+        user: &ManagedAddress,
+        week: Week,
+    ) -> SingleValueMapper<bool>;
+}
+
+#[cfg(test)]
+mod claim_progress_tests {
```

Comment:
Why have a test here?

Reply:
It's a test for the struct's base logic, so there's not much reason to bundle it with the general contract tests.